data/tutorials/guides/0tt_02_file_manipulation.md

```diff
@@ -8,123 +8,618 @@ category: "Tutorials"
+```
+
+This code gets the size of `file.txt` and prints it.
+
+### Working with File Paths
+
+OCaml's `Filename` module provides several functions to work with file paths.
+
+```ocaml
+let dir = Filename.dirname "/path/to/file.txt" in
+let base = Filename.basename "/path/to/file.txt" in
+let full = Filename.concat dir "new_file.txt" in
+print_endline dir;
+print_endline base;
+print_endline full;
+```
+
+This code extracts the directory and base name from a path, and it concatenates a
+directory name with a base name.
+
+### Creating Temporary Files and Directories
+
+The `Filename` module also provides functions to create temporary files and
+directories.
+
+```ocaml
+let temp_file = Filename.temp_file "prefix" ".suffix" in
+let temp_dir = Filename.temp_dir "prefix" in
+print_endline temp_file;
+print_endline temp_dir;
+```
+
+This code creates a temporary file and a temporary directory, then it prints their
+paths.
+
+## Error Handling
+
+As with any operation involving I/O, file operations in OCaml can fail for a
+variety of reasons, such as when a file does not exist, a directory cannot be
+created due to insufficient permissions, or when a disk is full. Handling such errors
+is crucial to writing robust and resilient programs. This section will cover
+some common error handling techniques in OCaml for file manipulation tasks.
+General purpose error handling in OCaml is also addressed in a
+[dedicated tutorial](/docs/error-handling)
+
+### Catching Exceptions
+
+In OCaml, many of the file and directory operations can raise an exception when
+they encounter an error. For example, trying to open a nonexistent file with
+`In_channel.open_text` will raise a `Sys_error` exception. Therefore, it's a
+good practice to catch these exceptions and handle them accordingly.
+
+Let's see how we can handle exceptions when reading from a file:
+
+```ocaml
+let read_from_file filename =
+  try
+    let ic = In_channel.open_text filename in
+    let content = In_channel.input_all ic in
+    In_channel.close ic;
+    content
+  with
+  | Sys_error msg -> print_endline ("Could not read file: " ^ msg); ""
+```
+
+In this function, we use a `try ... with` block to catch any `Sys_error` that
+might be raised when opening the file, reading its content, or closing it. If
+such an exception is raised, we print an error message and return an empty
+string.
+
+### Checking Beforehand
+
+Another way to avoid exceptions is to check beforehand whether an operation is
+likely to succeed. For example, before opening a file, you can check whether it
+exists and whether you have the necessary permissions to open it:
+
+```ocaml
+let read_from_file filename =
+  if Sys.file_exists filename then
+    let ic = In_channel.open_text filename in
+    let content = In_channel.input_all ic in
+    In_channel.close ic;
+    content
+  else
+    print_endline ("File does not exist: " ^ filename); ""
+```
+
+This function checks whether the file exists before trying to open it. If the
```

Comment:
This sentence repeats what's said above the example. If this wasn't intentional, perhaps delete it from 416.